app/src/main/java/com/arctouch/codechallenge/Controller.java

```diff
@@ -0,0 +1,126 @@
+                .client(new OkHttpClient.Builder().build())
+                .addConverterFactory(MoshiConverterFactory.create())
+                .addCallAdapterFactory(RxJava2CallAdapterFactory.create())
+                .build()
+                .create(TmdbApi.class);
+
+        genreDao = new GenreDao(this);
+        movieDao = new MovieDao(this);
+    }
+
+    public static synchronized Controller getInstance() {
+        return instance;
+    }
+
+    public void setMovieListReceiver (MovieListReceiver movieListReceiver){
+        this.movieListReceiver = movieListReceiver;
+    }
+
+    @SuppressLint("CheckResult")
+    public void getInitialData(String requestKey) {
+        if (networkObserver.isNetworkAvailable(this)) {
+            movieListReceiver.hideLog();
+
+            api.genres(TmdbApi.API_KEY, TmdbApi.DEFAULT_LANGUAGE)
+                    .subscribeOn(Schedulers.io())
+                    .observeOn(AndroidSchedulers.mainThread())
+                    .subscribe(response -> {
+                        Cache.setGenres(response.genres);
+                        genreDao.insertRegisterList(response.genres);
+                        getUpcomingMovies(1, requestKey);
+                    });
+        } else {
+            Cache.setGenres(genreDao.getAllRegisters());
+            Toast.makeText(instance, Cache.getGenres().get(0).name, Toast.LENGTH_SHORT).show();
+            movieListReceiver.displayLog(getResources().getStringArray(R.array.warning)[0]);
+        }
+    }
+
+    @SuppressLint("CheckResult")
+    public void getUpcomingMovies(Integer page, String requestKey){
+
+        if (networkObserver.isNetworkAvailable(this)) {
+            movieListReceiver.hideLog();
+            api.upcomingMovies(TmdbApi.API_KEY, TmdbApi.DEFAULT_LANGUAGE, Long.valueOf(page), TmdbApi.DEFAULT_REGION)
+                .subscribeOn(Schedulers.io())
+                .observeOn(AndroidSchedulers.mainThread())
+                .subscribe(response -> returnDataToList(response, requestKey));
+        } else
+            movieListReceiver.displayLog(getResources().getStringArray(R.array.warning)[0]);
+    }
+
+    @SuppressLint("CheckResult")
+    public void getMovieByQuery(String query, Integer page, String requestKey){
+
+        if (networkObserver.isNetworkAvailable(this)) {
+            movieListReceiver.hideLog();
+            api.searchMovies(TmdbApi.API_KEY, TmdbApi.DEFAULT_LANGUAGE, query, Long.valueOf(page), TmdbApi.DEFAULT_REGION)
+                .subscribeOn(Schedulers.io())
+                .observeOn(AndroidSchedulers.mainThread())
+                .subscribe(response -> returnDataToList(response, requestKey));
+        } else
+            movieListReceiver.displayLog(getResources().getStringArray(R.array.warning)[0]);
+    }
+
+    private void returnDataToList(UpcomingMoviesResponse response, String requestKey){
+        for (Movie movie : response.results) {
+            List<Genre> genreList = new ArrayList<>();
```

Comment:
Why did you use an `ArrayList` specifically? Are there alternative `List` implementations? What are the pros and cons here?